core/store/src/trie/mod.rs

```diff
@@ -734,6 +743,81 @@ impl Trie {
     pub fn get_trie_nodes_count(&self) -> TrieNodesCount {
         self.storage.get_trie_nodes_count()
     }
+
+    /// Given a root and a key, it returns a MerkleProof for that key.
+    /// Note that it can be a proof of membership, or a proof of non membership. This is encoded
+    /// in the boolean of the return type.
+    ///
+    /// Hence, if [get_proof](#method.get_proof) returns (true, _), it means that the key-value are found in the trie
+    /// and a proof for that is presented. Similarly, if (false, _) is returned, it means that the key-value
+    /// pair does not belong to the trie and a proof for that is also returned.
+    ///
+    /// # Errors
+    ///
+    /// If the root or any subsequent is not found in the trie, `StorageError` will be returned
+    pub fn get_proof(
+        &self,
+        root: &CryptoHash,
+        key: &[u8],
+    ) -> Result<(bool, Vec<RawTrieNodeWithSize>), StorageError> {
+        let mut key = NibbleSlice::new(key);
+        let mut hash = *root;
+        if hash == Trie::EMPTY_ROOT {
+            return Ok((key.is_empty(), Vec::new()));
+        }
+
+        let mut levels: Vec<RawTrieNodeWithSize> = vec![];
+        loop {
+            let bytes = self.storage.retrieve_raw_bytes(&hash)?;
+            let node = RawTrieNodeWithSize::decode(&bytes).map_err(|_| {
+                StorageError::StorageInconsistentState("RawTrieNode decode failed".to_string())
+            })?;
+
+            match node.node {
+                RawTrieNode::Leaf(existing_key, value_length, value_hash) => {
+                    let found = NibbleSlice::from_encoded(&existing_key).0 == key;
+                    levels.push(RawTrieNodeWithSize {
+                        node: RawTrieNode::Leaf(existing_key, value_length, value_hash),
+                        memory_usage: node.memory_usage,
+                    });
+                    return Ok((found, levels));
+                }
+                RawTrieNode::Extension(existing_key, child_hash) => {
+                    let existing_key_nibble = NibbleSlice::from_encoded(&existing_key).0;
+                    let found = key.starts_with(&existing_key_nibble);
+
+                    key = key.mid(existing_key_nibble.len());
+                    hash = child_hash;
+
+                    levels.push(RawTrieNodeWithSize {
+                        node: RawTrieNode::Extension(existing_key, child_hash),
+                        memory_usage: node.memory_usage,
+                    });
+                    if !found {
+                        return Ok((false, levels));
+                    }
```

Comment:
```suggestion
                    let existing_key_nibble = NibbleSlice::from_encoded(&existing_key).0;
                    levels.push(RawTrieNodeWithSize {
                        node: RawTrieNode::Extension(existing_key, child_hash),
                        memory_usage: node.memory_usage,
                    });
                    let found = key.starts_with(&existing_key_nibble);
                    if !found {
                        return Ok((false, levels));
                    }
                    key = key.mid(existing_key_nibble.len());
                    hash = child_hash;
```

I find this order of statements easier to read and also it might help the compiler in doing less unnecessary work.

Reply:
This will force you to clone `existing_key`, since it's taken as a reference by `existing_key_nibble` which is used afterwards

Reply:
Ah, got it!